tardis/model/parse_input.py

```diff
@@ -146,10 +248,235 @@ def parse_abundance_section(config, atom_data, geometry):
+        abundances_section = csvy_model_config.abundance
+        mass_fraction, isotope_mass_fraction = read_uniform_abundances(
+            abundances_section, geometry.no_of_shells
+        )
+    else:
+        index, mass_fraction, isotope_mass_fraction = parse_csv_abundances(
+            csvy_model_data
+        )
+        mass_fraction = mass_fraction.loc[:, 1:]
+        mass_fraction.columns = np.arange(mass_fraction.shape[1])
+        isotope_mass_fraction = isotope_mass_fraction.loc[:, 1:]
+        isotope_mass_fraction.columns = np.arange(
+            isotope_mass_fraction.shape[1]
+        )
+
+    mass_fraction = mass_fraction.replace(np.nan, 0.0)
+    mass_fraction = mass_fraction[mass_fraction.sum(axis=1) > 0]
+    isotope_mass_fraction = isotope_mass_fraction.replace(np.nan, 0.0)
+    isotope_mass_fraction = isotope_mass_fraction[
+        isotope_mass_fraction.sum(axis=1) > 0
+    ]
+    norm_factor = mass_fraction.sum(axis=0) + isotope_mass_fraction.sum(axis=0)
+
+    if np.any(np.abs(norm_factor - 1) > 1e-12):
+        logger.warning(
+            "Abundances have not been normalized to 1." " - normalizing"
+        )
+        mass_fraction /= norm_factor
+        isotope_mass_fraction /= norm_factor
+
+    isotope_mass_fraction = IsotopeAbundances(
+        isotope_mass_fraction, time_0=csvy_model_config.model_isotope_time_0
+    ).decay(time_explosion)
+    return convert_to_nuclide_mass_fraction(
+        isotope_mass_fraction, mass_fraction
+    )
+
+
+def parse_density_csvy(csvy_model_config, csvy_model_data, time_explosion):
+    """
+    Parse the density data from a CSVY model.
+
+    Parameters
+    ----------
+    csvy_model_config : object
+        The configuration data of the CSVY model.
+    csvy_model_data : object
+        The data of the CSVY model.
+    time_explosion : float
+        The time of the explosion.
+
+    Returns
+    -------
+    density : object
+        The parsed density data.
+
+    Raises
+    ------
+    None.
+
+    Notes
+    -----
+    This function parses the density data from a CSVY model. If the CSVY model configuration
+    contains a 'density' attribute, it uses the 'parse_csvy_density' function to parse the
+    density data. Otherwise, it calculates the density data using the 'calculate_density_after_time'
+    function. The parsed density data is returned.
+    """
+
+    if hasattr(csvy_model_config, "density"):
+        density = parse_csvy_density(csvy_model_config, time_explosion)
+    else:
+        time_0 = csvy_model_config.model_density_time_0
+        density_field_index = [
+            field["name"] for field in csvy_model_config.datatype.fields
+        ].index("density")
+        density_unit = u.Unit(
+            csvy_model_config.datatype.fields[density_field_index]["unit"]
+        )
+        density_0 = csvy_model_data["density"].values * density_unit
+        density_0 = density_0.to("g/cm^3")[1:]
+        density_0 = density_0.insert(0, 0)
```

Comment:
??

Reply:
See #2454

Reply:
just talked to @AlexHls - will do after this is merged.